File: src/main/java/com/example/TelegramBot/MyMemoryTranslator.java
package com.example.TelegramBot;

import java.net.URLEncoder;
import org.apache.http.client.methods.*;
import org.apache.http.impl.client.*;
import org.apache.http.util.EntityUtils;
import com.fasterxml.jackson.databind.ObjectMapper;

public class MyMemoryTranslator {
    private static final String API_URL = "https://api.mymemory.translated.net/get";
    private final ObjectMapper mapper = new ObjectMapper();

    public String translate(String text, String fromLang, String toLang) throws Exception {
        try {
            // Encode ALL special characters including text and language pair
            String encodedText = URLEncoder.encode(text, "UTF-8");
            String encodedLangPair = URLEncoder.encode(fromLang + "|" + toLang, "UTF-8");

            String url = String.format("%s?q=%s&langpair=%s",
                    API_URL, encodedText, encodedLangPair);

            System.out.println("Final URL: " + url); // Debug log

            try (CloseableHttpClient client = HttpClients.createDefault()) {
                HttpGet request = new HttpGet(url);

                try (CloseableHttpResponse response = client.execute(request)) {
                    String jsonResponse = EntityUtils.toString(response.getEntity());
                    return mapper.readTree(jsonResponse)
                            .path("responseData")
                            .path("translatedText")
                            .asText();
                }
            }
        } catch (Exception e) {
            System.err.println("Full error details:");
            e.printStackTrace();
            throw new Exception("Translation failed. Please try simpler text.");
        }
    }
}

File: my_memory_translator.py
"""
MyMemory Translator API client for translating text.
"""
import urllib.parse
import urllib.request
import json


class MyMemoryTranslator:
    """A class to handle translation using MyMemory API."""
    
    API_URL = "https://api.mymemory.translated.net/get"
    
    def translate(self, text: str, from_lang: str, to_lang: str) -> str:
        """
        Translate text from one language to another using MyMemory API.
        
        Args:
            text: The text to translate
            from_lang: Source language code (e.g., 'en')
            to_lang: Target language code (e.g., 'es')
            
        Returns:
            The translated text
            
        Raises:
            Exception: If translation fails
        """
        try:
            # Encode ALL special characters including text and language pair
            encoded_text = urllib.parse.quote(text, safe='')
            encoded_lang_pair = urllib.parse.quote(f"{from_lang}|{to_lang}", safe='')
            
            url = f"{self.API_URL}?q={encoded_text}&langpair={encoded_lang_pair}"
            
            print(f"Final URL: {url}")  # Debug log
            
            request = urllib.request.Request(url)
            with urllib.request.urlopen(request) as response:
                json_response = response.read().decode('utf-8')
                data = json.loads(json_response)
                
                translated_text = data.get('responseData', {}).get('translatedText', '')
                return translated_text
                
        except Exception as e:
            print("Full error details:")
            print(str(e))
            raise Exception("Translation failed. Please try simpler text.")
